skbio/stats/composition.py

```diff
@@ -2058,3 +2061,423 @@ def dirmult_ttest(
+    3        Y2  treatment  0.744093 -1.582636  3.401875  0.252057  0.687051
+
+    """
+
+    # Test if data is a pandas DataFrame
+    errmsg = (
+        "%s must be a pandas DataFrame or a numpy structured or rec array or "
+        "a dictionary."
+    )
+
+    if not isinstance(data, pd.DataFrame):
+        try:
+            data = pd.DataFrame(data)
+        except (TypeError, ValueError):
+            raise TypeError(errmsg % "Data")
+
+    if data.ndim != 2:
+        try:
+            data = pd.DataFrame(data, index=data.dtype.names)
+        except (TypeError, ValueError):
+            raise TypeError(errmsg % "Data")
+
+    if not hasattr(data, "__getitem__"):
+        try:
+            data = pd.DataFrame(data)
+        except (TypeError, ValueError):
+            raise TypeError(errmsg % "Data")
+
+    if not isinstance(metadata, pd.DataFrame):
+        try:
+            metadata = pd.DataFrame(metadata)
+        except (TypeError, ValueError):
+            raise TypeError(errmsg % "Metadata")
+
+    # Test if data has missing values
+    if data.isnull().values.any():
+        raise ValueError("Cannot handle missing values in data.")
+
+    # Test if metadata has missing values
+    if metadata.isnull().values.any():
+        raise ValueError("Cannot handle missing values in metadata.")
+
+    # Test if metadata and data have the same index, regardless of order
+    if not data.index.sort_values().equals(metadata.index.sort_values()):
+        raise ValueError("Data and metadata must have the same index.")
+
+    rng = get_rng(seed)
+
+    posterior = [
+        rng.dirichlet(data.values[i] + pseudocount) for i in range(data.shape[0])
+    ]
+
+    dir_table = pd.DataFrame(clr(posterior), index=data.index, columns=data.columns)
+
+    _call_res = _lme_call(
+        formula=formula,
+        table=dir_table,
+        metadata=metadata,
+        groups=groups,
+        reml=reml,
+        method=method,
+        fit_kwargs=fit_kwargs,
+        **kwargs,
+    )
+
+    res = _call_res[0]
+    list_of_vars = _call_res[2]
+
+    # Creating an empty dict to store sum of values (Using a DataFrame threw errors)
+    # uses a separate array for each covariate and each feature
+    # array index: 0: pvalue, 1: CI(2.5), 2: CI(97.5), 3: log2fc, 4: qvalue
+    res_dict = {}
+
+    for feature in data.columns:
+        res_ = res_dict[feature] = {}
+        for covar in list_of_vars:
+            res_[covar] = [0] * 5
+
+    for single_covar_data in res:
+        group = res_dict[single_covar_data["FeatureID"]][single_covar_data["Covariate"]]
+        for i, key in enumerate(("pvalue", "CI(2.5)", "CI(97.5)", "Log2(FC)")):
+            group[i] = single_covar_data[key]
+
+    for i in range(1, draws):
+        posterior = [
+            rng.dirichlet(data.values[i] + pseudocount) for i in range(data.shape[0])
+        ]
+        dir_table = pd.DataFrame(clr(posterior), index=data.index, columns=data.columns)
```

Comment:
Suggested change above could also apply here, and I think also removes the need to construct the `df`. I suspect the `clr` operation could be performed within the `apply` as well

Reply:
Unfortunately the CLR function does not work with apply. I'm not sure why, but I have never been able to get it to work.  We got it to work after transposing the dataframe, but not directly. See code block 13 here : https://github.com/scikit-bio/scikit-bio-tutorials/blob/main/06-marker-inference/06-marker-inference.ipynb

If we *really* want to have the apply function, we can do the following
```
dir_table = data.apply(lambda x: np.log(rng.dirichlet(data.values[i] + pseudocount)) - np.log(rng.dirichlet(data.values[i] + pseudocount)).mean())
```